src/yaml/from_yaml.cr

```diff
@@ -22,39 +22,86 @@ end
   location = pull.location
-  value = pull.read_scalar
-  if value.empty?
+  value = pull.read_plain_scalar
+  if YAML::NULL_VALUES.includes? value
     nil
   else
-    raise YAML::ParseException.new("Expected nil, not #{value}", *location)
+    raise YAML::ParseException.new("Expected null, not '#{value}'", *location)
   end
 end
 
 def Bool.new(pull : YAML::PullParser)
-  pull.read_scalar == "true"
+  location = pull.location
+  value = pull.read_plain_scalar
+  if YAML::TRUE_VALUES.includes? value
+    true
+  elsif YAML::FALSE_VALUES.includes? value
+    false
+  else
+    raise YAML::ParseException.new("Expected boolean, not '#{value}'", *location)
+  end
 end
 
-{% for type in %w(Int8 Int16 Int32 Int64 UInt8 UInt16 UInt32 UInt64) %}
-  def {{type.id}}.new(pull : YAML::PullParser)
+# TODO: Ideally, it may be beter to use `for type in Int::Primitive.union_types`
+# but it is currently broken due to: https://github.com/crystal-lang/crystal/issues/4301
+{% for bits in [8, 16, 32, 64] %}
+  {% for type in %w(Int UInt) %}
+    def {{type.id}}{{bits}}.new(pull : YAML::PullParser)
+      location = pull.location
+      value = pull.read_plain_scalar
+
+      number = {% if type == "UInt" %}
+        value.to_u{{bits}}?(underscore: false, prefix: true)
+      {% else %}
+        value.to_i{{bits}}?(underscore: false, prefix: true)
+      {% end %}
+      return number if number
+
+      raise YAML::ParseException.new("Invalid {{type.id}}{{bits}} number", *location)
+    end
+  {% end %}
+{% end %}
+
+{% for bits in [32, 64] %}
+  def Float{{bits}}.new(pull : YAML::PullParser)
     location = pull.location
-    begin
-      {{type.id}}.new(pull.read_scalar)
-    rescue ex
-      raise YAML::ParseException.new(ex.message.not_nil!, *location)
+    value = pull.read_plain_scalar
+    if float = value.to_f{{bits}}?
+      float
+    elsif YAML::INFINITY_VALUES.includes? value.lchop('+')
+      INFINITY
+    elsif value[0]? == '-' && YAML::INFINITY_VALUES.includes?(value.lchop('-'))
+      -INFINITY
+    elsif YAML::NAN_VALUES.includes? value
+      NAN
+    else
+      raise YAML::ParseException.new("Invalid Float{{bits}} number", *location)
     end
   end
 {% end %}
 
-def String.new(pull : YAML::PullParser)
-  pull.read_scalar
-end
-
-def Float32.new(pull : YAML::PullParser)
-  pull.read_scalar.to_f32
+# TODO: Implement a Time parser that supports all the YAML formats
+def Time.new(pull : YAML::PullParser)
+  location = pull.location
+  begin
+    value = pull.read_plain_scalar
+    Time::Format::ISO_8601_DATE_TIME.parse(value)
+  rescue ex : Time::Format::Error
+    raise YAML::ParseException.new("Could not parse time from '#{value}'", *location)
+  end
 end
 
-def Float64.new(pull : YAML::PullParser)
-  pull.read_scalar.to_f64
+def String.new(pull : YAML::PullParser)
+  location = pull.location
+  if pull.data.scalar.style == LibYAML::ScalarStyle::PLAIN
+    pull.read_plain_scalar.tap do |value|
+      if YAML.reserved_value?(value)
```

Comment:
I think this is wrong. Why can't I read `Null` or `~` as `nil` if it appears a plain in a YAML file. Here we are not guessing the type from the value (like in Ruby), we want to read a String.

Reply:
Because you are reading a string, `null` is a reserved value, as is `~`. `nil` does not appear in the YAML spec as a reserved value, so that would come back as `"nil"` in this case.